File: boj/DivideAndConquer/BOJ_4256.py
# 백준 4256 트리
# 22.01.21
# https://github.com/skyqnaqna/algorithm_study

import sys 
# from collections import deque, defaultdict
rd = sys.stdin.readline
sys.setrecursionlimit(50000)

def postorder(s, e, root):
	for i in range(s, e):
		if preorder[root] == inorder[i]:
			postorder(s, i, root + 1)
			postorder(i + 1, e, root + i - s + 1) # root의 오른쪽 자식 인덱스는 왼쪽 서브트리 노드의 수 + 1
			answer.append(str(preorder[root]))


t = int(rd())
for _ in range(t):
	n = int(rd())
	preorder = list(map(int, rd().split()))
	inorder = list(map(int, rd().split()))
	answer = []
	postorder(0, n, 0)
	print(' '.join(answer))

File: boj/DivideAndConquer/BOJ_4256.swift
/*
 백준 4256 트리
 22.01.21
 https://github.com/skyqnaqna/algorithm_study
 */

import Foundation

let t = Int(readLine()!)!

var answer = ""

for _ in 0 ..< t {
  let n = Int(readLine()!)!
  
  let preorder = readLine()!.split(separator: " ").map { Int(String($0))! }
  let inorder = readLine()!.split(separator: " ").map { Int(String($0))! }
  var postResult = [Int]()
  
  func postorder(_ s: Int, _ e: Int, _ root: Int) {
    for i in s ..< e {
      if preorder[root] == inorder[i] {
        postorder(s, i, root + 1)
        postorder(i + 1, e, root + i - s + 1) // root의 오른쪽 자식의 인덱스는 왼쪽 서브트리의 노드 수 + 1
        postResult.append(preorder[root])
      }
    }
  }
  
  postorder(0, n, 0)
  
  answer += postResult.map { String($0) }.joined(separator: " ") + "\n"
}

print(answer)
